File: WPF/MVVM/BindableBase.vb
Imports System.ComponentModel

Namespace Common.MVVM

    Public Class BindableBase
        Implements INotifyPropertyChanged

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Protected Function SetProperty(Of T)(ByRef Source As T, ByVal Value As T, <Runtime.CompilerServices.CallerMemberName()> Optional ByVal PropertyName As String = Nothing) As Boolean
            If Not Object.Equals(Source, Value) Then
                Source = Value
                Me.NotifyPropertyChanged(PropertyName)
                Return True
            End If

            Return False
        End Function

        Protected Overridable Sub OnPropertyChanged(ByVal E As PropertyChangedEventArgs)
            RaiseEvent PropertyChanged(Me, E)
        End Sub

        Protected Sub NotifyPropertyChanged(<Runtime.CompilerServices.CallerMemberName()> Optional ByVal PropertyName As String = Nothing)
            Me.OnPropertyChanged(New PropertyChangedEventArgs(PropertyName))
        End Sub

    End Class

End Namespace

File: WPF/MVVM/BindableBase.cs
using System.ComponentModel;

namespace Ks
{
    namespace Common.MVVM
    {
        public class BindableBase : INotifyPropertyChanged
        {
            public event PropertyChangedEventHandler PropertyChanged;

            protected bool SetProperty<T>(ref T Source, T Value, [System.Runtime.CompilerServices.CallerMemberName()] string PropertyName = null)
            {
                if (!object.Equals(Source, Value))
                {
                    Source = Value;
                    this.NotifyPropertyChanged(PropertyName);
                    return true;
                }

                return false;
            }

            protected virtual void OnPropertyChanged(PropertyChangedEventArgs E)
            {
                PropertyChanged?.Invoke(this, E);
            }

            protected void NotifyPropertyChanged([System.Runtime.CompilerServices.CallerMemberName()] string PropertyName = null)
            {
                this.OnPropertyChanged(new PropertyChangedEventArgs(PropertyName));
            }
        }
    }
}
